### PR TITLE
Make compiletest set cwd before running js tests

### PR DESCRIPTION
**Pull request by richard-uk1**
Proposed fix for #38800.

Not all tests pass yet - I will mention failures here once the test suite has finished.

### CODE REVIEW

**Comment by derekdreery** [comment]
Update: The following tests fail on run-pass:

```
    [run-pass] run-pass/asm-concat-src.rs
    [run-pass] run-pass/conditional-compile-arch.rs
    [run-pass] run-pass/extern-call-deep.rs
    [run-pass] run-pass/issue-27859.rs
    [run-pass] run-pass/out-of-stack.rs
```

**Comment by aidanhs** [comment]
Thanks for the PR @derekdreery! We'll check in now and again to make sure @brson or another reviewer gets to this soon.

**Comment by carols10cents** [comment]
ping @brson, welcome back from PTO! here's a PR for you to review! ❤️

**Comment by alexcrichton** [comment]
@derekdreery I'm forgetting why, but can you remind me why this is necessary to do on these tests (change the cwd)? Also, I wonder if we may wish to change the cwd for all tests instead of just emscripten/wasm ones?

**Comment by derekdreery** [comment]
The emscripten tools (node) expect the wasm file to be in the cwd, so tests don't work without setting the cwd to the directory of the file.

> I wonder if we may wish to change the cwd for all tests instead of just emscripten/wasm ones?

I seem to recall that @brson was worried that this would cause test failures for other targets, but I may be mistaken.

**Comment by alexcrichton** [comment]
Ah ok, can we try setting it for all tests perhaps? It'd be great to avoid a bunch of "if-wasm-else-this" statements throughout

**Comment by derekdreery** [comment]
Ok I'll have a go. I'll let travis do the work rather than run it locally.

**Comment by Mark-Simulacrum** [comment]
I think we're almost there but a codegen test needs fixing:

```
[00:48:17] ---- [codegen] codegen/remap_path_prefix/main.rs stdout ----
[00:48:17] 	
[00:48:17] error: verification with 'FileCheck' failed
[00:48:17] status: exit code: 1
[00:48:17] command: /usr/lib/llvm-3.7/bin/FileCheck -input-file=/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main.ll /checkout/src/test/codegen/remap_path_prefix/main.rs
[00:48:17] stdout:
[00:48:17] ------------------------------------------
[00:48:17] 
[00:48:17] ------------------------------------------
[00:48:17] stderr:
[00:48:17] ------------------------------------------
[00:48:17] /checkout/src/test/codegen/remap_path_prefix/main.rs:28:11: error: expected string not found in input
[00:48:17] // CHECK: !DIFile(filename: "/the/src/remap_path_prefix/main.rs", directory: "/the/cwd")
[00:48:17]           ^
[00:48:17] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main.ll:9:1: note: scanning from here
[00:48:17] @__rustc_debug_gdb_scripts_section__ = internal unnamed_addr constant [34 x i8] c"\01gdb_load_rust_pretty_printers.py\00", section ".debug_gdb_scripts", align 1
[00:48:17] ^
[00:48:17] /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix/main.ll:44:6: note: possible intended match here
[00:48:17] !1 = !DIFile(filename: "/the/src/remap_path_prefix/main.rs", directory: "/the/cwd/build/x86_64-unknown-linux-gnu/test/codegen/remap_path_prefix")
[00:48:17]      ^
[00:48:17] 
[00:48:17] ------------------------------------------
[00:48:17] 
[00:48:17] thread '[codegen] codegen/remap_path_prefix/main.rs' panicked at 'explicit panic', /checkout/src/tools/compiletest/src/runtest.rs:2480
[00:48:17] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:48:17] 
[00:48:17] 
[00:48:17] failures:
[00:48:17]     [codegen] codegen/remap_path_prefix/main.rs
```

**Comment by arielb1** [comment]
@derekdreery

This looks ready except for the `remap_path_prefix/main.rs
` test. Can you fix it so we can merge this?

File: src/librustc_back/target/wasm32_unknown_emscripten.rs [hunk 1]
```diff
@@ -18,6 +18,8 @@ pub fn target() -> Result<Target, String> {
                           vec!["-s".to_string(),
                                "BINARYEN=1".to_string(),
                                "-s".to_string(),
+                               "BINARYEN_METHOD='native-wasm,interpret-binary'".to_string(),
```
Comment (malbarbo) [inline src/librustc_back/target/wasm32_unknown_emscripten.rs]: interpret-binary is not necessary any more, node 8.0 can run wasm. See https://github.com/rust-lang/rust/pull/42631

Comment (derekdreery) [inline src/librustc_back/target/wasm32_unknown_emscripten.rs]: ooh cool I'll update

**Comment by derekdreery** [comment]
@arielb1 I'm not familiar with the LLVM instructions in that test - can you give me some pointers?

**Comment by alexcrichton** [comment]
I think that's because the cwd of rustc itself is changed, right? Perhaps the cwd could only be changed for the test process and not the rustc process?

**Comment by derekdreery** [comment]
`std::process` is used to spawn the test commands. The current_dir method on the builder is used to set the cwd. I assume this means that the cwd is set just for that command, not the container command.

**Comment by alexcrichton** [comment]
Indeed yeah! And for the commands just for the target process, not for rustc itself.

**Comment by alexcrichton** [comment]
ping @derekdreery, just wanted to keep this on your radar!

**Comment by derekdreery** [comment]
Hi I'm just unsure of how to proceed.

**Comment by alexcrichton** [comment]
@derekdreery does my [comment above](https://github.com/rust-lang/rust/pull/42059#issuecomment-308444695) make sense? Only setting the cwd for the test process, not the rustc one?

**Comment by shepmaster** [comment]
@derekdreery just checking in to make sure you saw @alexcrichton's [earlier comment](https://github.com/rust-lang/rust/pull/42059#issuecomment-308444695). Please let us know if there's more information we can provide to help.

**Comment by derekdreery** [comment]
Hi sorry for the lead time. I've been away from my computer. Will endeavor
to fix this this week.

**Comment by arielb1** [comment]
Hi @derekdreery - are you still looking on this PR? Just a friendly ping to make sure this isn't getting lost.

**Comment by derekdreery** [comment]
Apologies for the time. I made this PR because I've used rust quite a lot and want to understand/contribute to the internals a bit more. I'm finding doing this PR very helpful for getting used to the structure of rust-lang/rust, but it means my work is quite slow (not helped by having to earn money by doing paid work etc. :P )

**Comment by derekdreery** [comment]
OK so I've done some more work on this. Since node 8 supports wasm native (pointed out by @malbarbo), I used my own `node` rather than emsdk's. With version 8, it seems the *cwd* requirement has gone away! I'm able to run (most) of the `run-pass` tests fine (without my code changes). However, none of the `run-fail` tests work, claiming that their wasm files are missing (they are missing - I checked).

So I suspect that this PR will become redundant, but then there is the question: What is the best way to set up testing in CI? Is it to just override the node version used? And if so how do I get the `run-fail` tests working?

What do people think of this?

**Comment by alexcrichton** [comment]
@derekdreery oh the CI testing is all done through docker images in `src/ci/docker`, so it's just whatever we configure there. Do you have example for failures with run-fail tests?

**Comment by derekdreery** [comment]
I've changed `compilertest` to only change the working directory for the test process, and tests seem to run correctly (including stage 0/1 i.e. x64 -> x64 builds). Unfortunately, I'm using a custom script to remove the node path added by emscripten, so I can get native wasm in my node (v8.1.4):

```python
#!/usr/bin/python
import os

REMOVE_STRING = ":/home/rdodd/packages/emsdk-portable/node/4.1.1_64bit/bin"

def main():
    old_path = os.environ["PATH"]
    new_path = old_path.replace(REMOVE_STRING, "")
    print(new_path)

if __name__ == '__main__':
    main()
```

I could add some code to compiletest to do this (actually splitting the path, not using this hacky string replace), but it feels like it should be in the docker shell script. I'm going to let the automated tests run to see if this is indeed an issue, and then have a think about how to solve it.

**Comment by derekdreery** [comment]
We could use emulated wasm on node v4, but it would be an order of magnitude slower

**Comment by alexcrichton** [comment]
@derekdreery thanks! Do you think this is ready to merge? Or because you need a local workaround do you think we should hold off?

**Comment by derekdreery** [comment]
I think yes it can be merged, and the workaround can be added in a separate PR, if you want to run the test suite against emscripten target in the future. That test would fail currently as there are some things (e.g. inline asm) that won't work with emscripten, and those tests need to be ignored for this target.~

I'm confused as to why some mac targets are failing.

**Comment by aidanhs** [comment]
The logs were AWOL and those targets would be skipped anyway given it was just a PR build, so restarted the jobs and now they show as fine.

**Comment by alexcrichton** [comment]
I think with this implemented we should be able to [remove this workaround](https://github.com/rust-lang/rust/blob/master/src/ci/docker/disabled/wasm32/Dockerfile#L22), right? Could this PR also delete those shim `node.sh` scripts and ensure that the container still passes tests?

**Comment by arielb1** [comment]
> I think with this implemented we should be able to [remove this workaround](https://github.com/rust-lang/rust/blob/master/src/ci/docker/disabled/wasm32/Dockerfile#L22), right? Could this PR also delete those shim `node.sh` scripts and ensure that the container still passes tests?

@derekdreery - @alexcrichton thinks you can remove the workaround. Just a friendly ping to keep this on your radar.

**Comment by derekdreery** [comment]
Hi yeah sorry didn't realise you were waiting on me. I'll have a look.

**Comment by derekdreery** [comment]
I've removed the workaround you mention, and put node in the PATH.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Pull request merged**